src/odemis/driver/piezomotor.py

```diff
@@ -0,0 +1,681 @@
+                    char = self._serial.read()
+                except IOError:
+                    logging.warn("Failed to read from PMT Control firmware, "
+                                 "trying to reconnect.")
+                    if self._recovering:
+                        raise
+                    else:
+                        self._tryRecover()
+                        # don't send command again
+                        raise IOError("Failed to read from PMT Control firmware, "
+                                      "restarted serial connection.")
+                if not char:
+                    raise IOError("Timeout after receiving %s" % to_str_escape(resp))
+                else:
+                    resp += char
+            logging.debug("Received response %s", to_str_escape(resp))
+
+            # Check response (command should be echoed back)
+            if not resp.startswith(cmd[:-len(EOL)]):
+                raise ValueError("Response starts with %s != %s", resp[:len(cmd)], cmd)
+            if b"__??__" in resp or b"!" in resp:
+                raise ValueError("Received response %s, command %s not understood." % (resp, cmd))
+
+            # Format:
+            #    * for query with response: <cmd>:<ret><EOL> (will return <ret>)
+            #    * for set command without response: <cmd><EOL> (will return b"")
+            return resp[len(cmd) + 1 - len(EOL):-len(EOL)].decode("latin1")
+
+    def _tryRecover(self):
+        self._recovering = True
+        self.state._set_value(HwError("USB connection lost"), force_write=True)
+        # Retry to open the serial port (in case it was unplugged)
+        # _ser_access should already be acquired, but since it's an RLock it can be acquired
+        # again in the same thread
+        try:
+            with self._ser_access:
+                while True:
+                    try:
+                        self._serial.close()
+                        self._serial = None
+                    except Exception:
+                        pass
+                    try:
+                        logging.debug("Searching for the device on port %s", self._portpattern)
+                        self._port = self._findDevice(self._portpattern)
+                    except IOError:
+                        time.sleep(2)
+                    except Exception:
+                        logging.exception("Unexpected error while trying to recover device")
+                        raise
+                    else:
+                        # We found it back!
+                        break
+            # it now should be accessible again
+            self.state._set_value(model.ST_RUNNING, force_write=True)
+            logging.info("Recovered device on port %s", self._port)
+        finally:
+            self._recovering = False
+
+    def terminate(self):
+        # called in __del__ (see HwComponent)
+        self.setWaveform(WAVEFORM_PARK)  # power off
+        self._serial.close()
```

Comment:
Maybe it is nice to add a logging.info message holding "succefully terminated connection with the sonic stage driver"?